src/NuGet.Core/NuGet.DependencyResolver.Core/GraphModel/GraphNode.cs

```diff
@@ -10,14 +10,34 @@ namespace NuGet.DependencyResolver
 {
     public class GraphNode<TItem>
     {
-        public GraphNode(LibraryRange key)
+        public GraphNode(LibraryRange key, bool hasEmptyInnerNodes = false, bool hasEmptyParentNodes = false)
         {
             Key = key;
-            InnerNodes = new List<GraphNode<TItem>>();
             Disposition = Disposition.Acceptable;
-            ParentNodes = new List<GraphNode<TItem>>();
+
+            //Create nonEmpty InnerNodes only when it's nessecery (InnerNodes is empty when it has no dependencies, including runtime dependencies).
+            if (hasEmptyInnerNodes)
+            {
+                InnerNodes = EmptyList;
+            }
+            else
+            {
+                InnerNodes = new List<GraphNode<TItem>>();
+            }
+
+            //Create nonEmpty ParentNodes only when it's nessecery (ParentNodes is nonEmpty only for certain nodes when Central Package Management is enabled).
```

Comment:
```suggestion
            // ParentNodes is non-empty only for certain nodes when Central Package Management is enabled.
```

Reply:
Thanks! Since this comment is redundant, I removed them.